feat(xai): Add comprehensive tests and docs for agentic tool calling API

KushagraaWadhwa opened the pull request:
## Description

Adds comprehensive test coverage and documentation for X.AI's agentic tool calling API support in `langchain-xai`.

X.AI is deprecating the Live Search API by **December 15, 2025** in favor of the new agentic tool calling API ([docs](https://docs.x.ai/docs/guides/tools/overview)). While the `server_tools` parameter already provides support for this feature, it lacked test coverage and documentation visibility, leading to user confusion.

**Issue:** Closes #33961

## Changes

### 🧪 Test Coverage Added

**Unit Tests** (`tests/unit_tests/test_chat_models.py`):
- `test_server_tools_param()`: Validates server_tools parameter handling
- `test_server_tools_and_search_parameters_conflict()`: Ensures both APIs can't be used together  
- `test_search_parameters_deprecation_warning()`: Verifies deprecation warnings

**Integration Tests** (`tests/integration_tests/test_chat_models.py`):
- `test_server_tools_web_search()`: Tests autonomous web search
- `test_server_tools_x_search()`: Tests X (Twitter) search
- `test_server_tools_code_execution()`: Tests code execution capability
- `test_server_tools_multiple()`: Tests using multiple tools simultaneously

### 📚 Documentation Enhancements

- Enhanced `ChatXAI` docstring with comprehensive agentic tool calling examples
- Added deprecation notices for `search_parameters` API
- Added validation to prevent using both old and new APIs simultaneously
- Added deprecation warnings with clear migration path
- Updated README with migration notice and quick examples

### 🔧 Code Quality

- Fixed import linting errors (PLC0415, I001)
- Properly sorted imports into standard groups

## Migration Example

```python
# Old (deprecated - remove by Dec 15, 2025)
model = ChatXAI(
    model="grok-4",
    search_parameters={"mode": "auto"},
)

# New (recommended)
model = ChatXAI(
    model="grok-4",
    server_tools=[{"type": "web_search"}],
)

# Multiple tools
model = ChatXAI(
    model="grok-4",
    server_tools=[
        {"type": "web_search"},
        {"type": "x_search"},
        {"type": "code_execution"},
    ],
)
```

## Available Server Tools

- **`web_search`**: Autonomous web search for current information
- **`x_search`**: Search posts on X (formerly Twitter)
- **`code_execution`**: Execute code to solve complex problems

## Dependencies

None - no new dependencies added.

## Testing

```bash
cd libs/partners/xai
make format  # ✅ Passed
make lint    # ✅ Passed
make test    # ✅ Passed
```

All tests pass locally and are backwards compatible.

## Breaking Changes

**None** - this PR is fully backward compatible. The old `search_parameters` API continues to work with deprecation warnings.

Commit:
feat(xai): Add comprehensive tests and docs for agentic tool calling API

- Add unit tests for server_tools parameter validation
- Add integration tests for web_search, x_search, and code_execution
- Enhance ChatXAI docstring with agentic tools examples
- Add deprecation warnings for search_parameters API
- Add validation to prevent using both old and new APIs
- Update README with migration notice

The server_tools parameter enables X.AI's new agentic tool calling API,
replacing the deprecated search_parameters (sunset Dec 15, 2025).

Available tools:
- web_search: Autonomous web search
- x_search: Search X (Twitter) posts
- code_execution: Execute code to solve problems

Commit:
fix(xai): Move warnings import to top-level to resolve linting error

PLC0415: Import statements must be at module level, not inside functions.
Moved 'import warnings' from validate_environment method to top of file.

Commit:
style(xai): Sort imports to fix I001 linting error

Reorder imports into proper groups:
1. __future__ imports
2. Standard library (warnings, typing)
3. Third-party (openai, pydantic, typing_extensions)
4. Local packages (langchain_core, langchain_openai)

Fixes ruff I001: Import block is un-sorted or un-formatted

KushagraaWadhwa committed:
Merge branch 'master' into feat/xai-agentic-tools-clean

Commit:
fix(xai): Fix import sorting to satisfy ruff I001

Alphabetize third-party imports (langchain_core before pydantic)
to match the repository's import ordering rules.

Commit:
fix(xai): Fix mypy type errors in tests

- Add None check before len() on server_tools (can be Optional)
- Handle response.content as str or list for .lower() call

Fixes mypy errors:
- tests/unit_tests/test_chat_models.py:149 (arg-type)
- tests/integration_tests/test_chat_models.py:149 (union-attr)

Commit:
style(xai): Fix line length in integration test

Break long line into multi-line expression to satisfy E501 (line <= 88 chars)

Commit:
style(xai): Apply ruff formatter to integration test

Use formatter-preferred single-line ternary within parentheses

mdrxy commented on libs/partners/xai/tests/integration_tests/test_chat_models.py:
```diff
@@ -95,3 +96,85 @@ def test_web_search() -> None:
+
+
+def test_server_tools_x_search() -> None:
+    """Test new agentic tool calling API with x_search."""
```
```suggestion
    """Test agentic tool calling API with x_search."""
```

mdrxy commented on libs/partners/xai/langchain_xai/chat_models.py:
```diff
@@ -408,8 +447,36 @@ class Joke(BaseModel):
+    """Parameters for search requests. Example: `{"mode": "auto"}`.
+
+    !!! warning "Deprecated"
+        The Live Search API will be deprecated by December 15, 2025.
```
```suggestion

        The Live Search API will be deprecated by December 15, 2025.
```

KushagraaWadhwa committed:
Update libs/partners/xai/README.md

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/tests/integration_tests/test_chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/tests/integration_tests/test_chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/tests/integration_tests/test_chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/tests/integration_tests/test_chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/langchain_xai/chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Update libs/partners/xai/langchain_xai/chat_models.py

Co-authored-by: Mason Daugherty <github@mdrxy.com>

KushagraaWadhwa committed:
Merge branch 'master' into feat/xai-agentic-tools-clean

Commit:
docs(xai): Remove references to non-existent files in README

Remove links to MIGRATION_GUIDE.md and examples/agentic_tools_example.py
that don't exist in this PR. Link to X.AI docs instead.

Addresses reviewer feedback.

KushagraaWadhwa committed:
Merge branch 'master' into feat/xai-agentic-tools-clean

Ce1ling commented:
Any updates on this PR?
